challenges/week1.longest-substring-with-repetition/puchi.js

```diff
@@ -0,0 +1,57 @@
+// v2
+// Runtime: 88 ms, faster than 78.97% of JavaScript online submissions for Longest Substring with At Least K Repeating Characters.
+// Memory Usage: 40.1 MB, less than 77.25% of JavaScript online submissions for Longest Substring with At Least K Repeating Characters.
+const longestSubstring = (str, k) => {
+  if (k < 2) return str.length;
+
+  let maxLength = 0;
+
+  const find = substr => {
+    // count repetitions
+    const seen = {};
+    for (const char of substr) seen[char] = seen[char] + 1 || 1;
+    // find characters that repeted less than `k` times
+    const less = Object.keys(seen).filter(char => seen[char] < k);
+    // update max length if the condition is met
+    if (!less.length) {
+      maxLength = Math.max(maxLength, substr.length);
+      return;
+    }
```

Comment:
기껏 recursive하게 코드를 짰으니깐...`maxLength` 를 업데이트하는 대신 `substr.length`를 리턴하도록 만들면 좀 더 이뻐 보일 것이라고 생각합니다. 물론 반대급부로 밑에서 `find`의 리턴값을 처리해야하는 나쁜 면이 있겠지만...
```suggestion
    if (!less.length) return substr.length;
```

Reply:
요즘 고민하던 건데 퓨어한 함수를 만들 것인가 리턴 포인트를 줄일 것인가.. 라는 문제. 어떻게 생각하시는지

Reply:
사실 위 변경전 코드가 다른 언어에서는 성능상으로는 미묘하게 이득일 수도 있는데 (마지막 호출 시 메모리 스택에 호출 목록을 저장 안해도 되므로) 거의 차이가 없을 것이므로 사실상 취향의 문제라고 생각함.

다만 recursive 함수는 그 특성상 디버깅 같은게 복잡해질 가능성이 있으므로 퓨어한 함수를 만들면 테스트에 조금이나마 더 유리하지 않나 싶습니다.

이거랑 별개로 위 코멘트의 "기껏 recursive하게 코드를 짰으니깐..." 의 의미는, 전통적으로 recursive 함수는 1개의 리턴 값을 갖도록 만들어 왔으므로 (수학의 `y = f(x)` 에서 비롯된 것이므로) 바꿔보면 어떻겠냐 하는, 걍 틀딱같은 소리를 한것임미다.

Reply:
재귀를 사용했던 건 정해지지 않은 횟수만큼 브랜치를 만들려고 했던건데, 좋은 코딩 습관을 들일 겸 근본에 충실한 게 맞는 거 같아여.. 아니면 재귀를 하지 않든가 ㅋㅋㅋ 살짝 고쳐서 올려볼게요

Reply:
재귀 함수가 리턴값을 갖도록 고쳤는데, 공간 효율이 91.85%까지 올랐습니다.. 뭐 단순히 Array.prototype.reduce가 for...of 반복보다 빠른 점도 있지만, 글쎄요, 재귀 호출이 발생할 때마다 클로저에 의해 캡쳐된 데이터가 적어서 그럴지도?